### PR TITLE
Adds a task for HA etcd kubeadm

### PR DESCRIPTION
**Pull request by chuckha**
Signed-off-by: Chuck Ha <ha.chuck@gmail.com>

### CODE REVIEW

**Comment by chuckha** [comment]
also cc @Bradamant3 (I'll take care of the hugo transition later, I want to get some eyes on this for the time being)

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+```
+cat << EOF > kubeadmcfg.yaml
+apiVersion: "kubeadm.k8s.io/v1alpha1"
+kind: MasterConfiguration
+etcd:
+    serverCertSANs:
+    - "${HOST1}"
+    - "${HOST2}"
+    - "${HOST3}"
```
Comment (detiber) [inline docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: Since you are generating unique certs on each etcd host using the copied ca cert, this should probably contain ${CURRENT_HOST} only.

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    peerCertSANs:
+    - "${HOST1}"
+    - "${HOST2}"
+    - "${HOST3}"
```
Comment (detiber) [inline docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: Same here.

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        advertise-client-urls: https://${CURRENT_HOST}:2379
+        initial-advertise-peer-urls: https://${CURRENT_HOST}:2380
+EOF
+```
+
+## Certificates
+
+### Certificate Authority
+
+Create a CA for the etcd certs. This only has to be done on one node.
```
Comment (chulkilee) [inline docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: Is it possible to use CA generated outside of kubeadm? It looks like if `/etc/kubernetes/pki/etcd/ca.{crt,key}` exists, kubeadm does not try to create it. Then could we mention that too?

I think this would be useful when using other tools with kubeadm

Comment (chuckha) [reply docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: I think that's worth mentioning. I'll add some copy. Thanks for the suggestion.

File: _data/tasks.yml [hunk 4]
```diff
@@ -166,7 +166,7 @@ toc:
   - docs/tasks/administer-cluster/reserve-compute-resources.md
   - docs/tasks/administer-cluster/guaranteed-scheduling-critical-addon-pods.md
   - docs/tasks/administer-cluster/declare-network-policy.md
-  - docs/tasks/administer-cluster/kms-provider.md      
+  - docs/tasks/administer-cluster/kms-provider.md
```
Comment (chuckha) [inline _data/tasks.yml]: editor automation 😪

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+through the process of creating a highly available etcd cluster of three members
+that can be used as an external etcd when using kubeadm to set up a kubernetes
+cluster.
+{% endcapture %}
+
+{% capture prerequisites %}
+* Three hosts that can talk to each other over default etcd ports (2379 for
+  clients and 2380 for peers).
+* The `kubeadm` binary installed on each host.
+* SSH or similar to one host and from that host, SSH needs to work to both other
```
Comment (jamiehannaford) [inline docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: Could we clarify the wording here?

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+  hosts (to copy certificate files around).
+* Docker installed on every host.
+* The `kubelet` installed and running with configuration to read static
```
Comment (jamiehannaford) [inline docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: maybe we could link them to our installation guide?

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+Note: You will have to define `HOST1`, `HOST2`, and `HOST3`. Examples could be
+`$ export HOST1=10.0.0.1` or a resolvable DNS name.
+
+Pay attention to the name parameter in extraArgs. It needs to be unique per
+host. This example uses infra0, infra1, and infra2, but it can be any valid etcd
+member name just so long as the names all match in the `initial-cluster`
+section.
+
+```
+export HOST1= # example: 10.0.0.1
```
Comment (jamiehannaford) [inline docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: could we use dummy vals for this? I think it helps users get an idea of what they need to replace with?

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+```
+kubeadm alpha phase certs etcd-ca
+```
+
+This creates two files
+
+1. `/etc/kubernetes/pki/etcd/ca.crt`
+2. `/etc/kubernetes/pki/etcd/ca.key`
+
+Copy these certs to the other two nodes and put them in `/etc/kubernetes/pki/etcd/`.
```
Comment (jamiehannaford) [inline docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: do you think it's a good idea to output the commands people can run so they can just copy/paste? I find it lowers the cognitive burden for beginners

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+{% capture overview %}
+Kubeadm by default will create a self-hosted etcd cluster of one member. This is
+not a great default for a highly available control plane. This task walks
+through the process of creating a highly available etcd cluster of three members
+that can be used as an external etcd when using kubeadm to set up a kubernetes
+cluster.
+{% endcapture %}
+
+{% capture prerequisites %}
+* Three hosts that can talk to each other over default etcd ports (2379 for
```
Comment (chulkilee) [inline docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: etcd ports can be configured via `kubeadmcfg.yaml` so it would be better if this wording reflects that.

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md (hunk 7 above)
Comment (chuckha) [reply docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: sure thing

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md (hunk 6 above)
Comment (chuckha) [reply docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: good call 👍

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md (hunk 5 above)
Comment (chuckha) [reply docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: Yeah, I struggled with this one and it shows. Basically, I want to convey that the users can copy files from host A to hosts B and C. SCP would be the easiest from a command line perspective...how about something like "SSH between hosts"? The only thing is that this is not really a requirement for the system itself and just to copy certs around. I'll mull on this for a little bit. Happy to listen to suggestions.

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md (hunk 8 above)
Comment (chuckha) [reply docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: I'm going to provide commands for users using ubuntu and we can discuss how you think that looks.

**Comment by chuckha** [comment]
@timothysc @jamiehannaford Changes in, would love some feedback, especially on the swizzling bit.

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md [hunk 10]
```diff
@@ -0,0 +1,143 @@
+---
+title: Setup a Highly Available Etcd Cluster With Kubeadm
+---
+
+{% capture overview %}
+Kubeadm by default will create a self-hosted etcd cluster of one member. This is
```
Comment (luxas) [inline docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: It's not self-hosted as per the definition of "self-hosting" we have in the community. I'd say it like this:

```
By default, kubeadm will run one etcd instance in a Static Pod ran by the kubelet on the master. This is not a highly-available setup in case your master with that one etcd replica goes unavailable.
```

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+that can be used as an external etcd when using kubeadm to set up a kubernetes
+cluster.
+{% endcapture %}
+
+{% capture prerequisites %}
+* Three hosts that can talk to each other over ports 2379 and 2380. This
+  document assumes these default ports. However, they are configurable through
+  the kubeadm config file.
+* Each host must [have docker, kubelet, and kubeadm installed][toolbox].
+* SSH between hosts is required for certificate copying.
```
Comment (luxas) [inline docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: Could you say you need _some_ infrastructure to copy files around instead of saying SSH is the only way to move files around?

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+cat << EOF > kubeadmcfg.yaml
+apiVersion: "kubeadm.k8s.io/v1alpha1"
+kind: MasterConfiguration
+etcd:
+    serverCertSANs:
+    - "${CURRENT_HOST}"
+    peerCertSANs:
+    - "${CURRENT_HOST}"
+    extraArgs:
+        initial-cluster-token: etcd-cluster-1
```
Comment (luxas) [inline docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: What is this used for?

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+  the kubeadm config file.
+* Each host must [have docker, kubelet, and kubeadm installed][toolbox].
+* SSH between hosts is required for certificate copying.
+
+[toolbox]: /docs/setup/independent/install-kubeadm/
+{% endcapture %}
+
+{% capture steps %}
+
+## Kubeadm Configuration File
```
Comment (luxas) [inline docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: Create a Configuration File for kubeadm to consume

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+
+Create a kubeadm configuration file on each host that looks something like this:
```
Comment (luxas) [inline docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: `something like what's scripted below.`?

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+        initial-cluster-state: new
+        name: ${NAME}
+        listen-peer-urls: https://${CURRENT_HOST}:2380
+        listen-client-urls: https://${CURRENT_HOST}:2379
+        advertise-client-urls: https://${CURRENT_HOST}:2379
+        initial-advertise-peer-urls: https://${CURRENT_HOST}:2380
+EOF
+```
+
+## Certificates
```
Comment (luxas) [inline docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: Generate certificates needed for the etcd cluster using the `kubeadm phase` command

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+
+### Certificate Authority
+
+If you already have a CA then the only action that is required is copying the
+`crt` and `key` file to `/etc/kubernetes/pki/etcd/ca.crt` and
+`/etc/kubernetes/pki/etcd/ca.key`. After those files have been copied, please
+skip to the Certificate Swizzling section below.
+
+Create a CA for the etcd certs. This only has to be done on one node.
```
Comment (luxas) [inline docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: s/one node/the first node/ maybe

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+        initial-advertise-peer-urls: https://${CURRENT_HOST}:2380
+EOF
+```
+
+## Certificates
+
+### Certificate Authority
+
+If you already have a CA then the only action that is required is copying the
+`crt` and `key` file to `/etc/kubernetes/pki/etcd/ca.crt` and
```
Comment (luxas) [inline docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: In the end, could you have a reference folder structure tree with a description of what every file should be named and is expected to contain?

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+
+### Certificate Swizzling
+
+Copy these certs to the other two nodes and put them in `/etc/kubernetes/pki/etcd/`.
+
+The steps to get these files on `$HOST2` might look like this:
+
+```
+root@HOST1 $ scp /etc/kubernetes/pki/etcd/ca.crt ubuntu@${HOST2}:/home/ubuntu/
+root@HOST1 $ scp /etc/kubernetes/pki/etcd/ca.key ubuntu@${HOST2}:/home/ubuntu/
```
Comment (luxas) [inline docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: Is the ca key needed for running an etcd peer on every master? Otherwise you might want to point out that you can also pregenerate everything and only distribute the unique client/serving certs to each master as well as the ca cert (only)

=> For security-concious folks

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+ubuntu@HOST2 $ sudo -s
+root@HOST2 $ chown root:root ca.crt && mv ca.crt /etc/kubernetes/pki/etcd/ca.crt
+root@HOST2 $ chown root:root ca.key && mv ca.key /etc/kubernetes/pki/etcd/ca.key
+```
+
+### Server Certs
```
Comment (luxas) [inline docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: More often referred to as "Serving Certs"? IDK, just what I've heard more often...

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+
+With the CA and Kubeadm configuration file in place on each host we can now
+create the certs.
+
+On each host run these kubeadm commands:
+
+```
+$ kubeadm alpha phase certs etcd-server --config=kubeadmcfg.yaml
+$ kubeadm alpha phase certs etcd-peer --config=kubeadmcfg.yaml
+$ kubeadm alpha phase certs etcd-healthcheck-client --config=kubeadmcfg.yaml
```
Comment (luxas) [inline docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: If this is a client cert, I'd rephrase the title to be "unique certs to be generated per peer" or similar

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+```
+
+## Manifests
+
+Now that the certs are in place we can create the manifest. On each host run the
+`kubeadm` command to generate a static manifest for etcd.
+
+```
+kubeadm alpha phase etcd local --config=kubeadmcfg.yaml
```
Comment (luxas) [inline docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: Every time I see alpha here I really look forward to the phases graduation :smile:

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+```
+
+## Optional: Check the cluster health
+
+### Get `etcdctl`
```
Comment (luxas) [inline docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: Could we use the docker image here instead so we don't have to curl/wget etc. instead just
```
docker run -it --net host -v /etc/kubernetes:/etc/kubernetes quay.io/coreos/etcd:v3.2.14 etcdctl
```

makes sense?

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+
+```
+wget --no-verbose https://github.com/coreos/etcd/releases/download/v3.1.13/etcd-v3.1.13-linux-amd64.tar.gz
+tar -xzf etcd-v3.13-linux-amd64.tar.gz
+```
+
+### Check the health
+
+```
+./etcd-v3.1.13-linux-amd64/etcdctl --cert-file /etc/kubernetes/pki/etcd/peer.crt --key-file /etc/kubernetes/pki/etcd/peer.key --ca-file /etc/kubernetes/pki/etcd/ca.crt --endpoints https://${HOST1}:2379 cluster-health
```
Comment (luxas) [inline docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: show successful output?

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+```
+{% endcapture %}
+
+{% capture discussion %}
+{% endcapture %}
+
+{% capture whatsnext %}
+* Learn how to use this HA etcd cluster as an external etcd  when bootstrapping
```
Comment (luxas) [inline docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: Add link to the next chapter here?

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md (hunk 12 above)
Comment (chuckha) [reply docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: Ah, great catch. We don't even need this line as it has a default value and we are not spinning up multiple clusters. Here is the relevant documentation:

> If spinning up multiple clusters (or creating and destroying a single cluster) with same configuration for testing purpose, it is highly recommended that each cluster is given a unique initial-cluster-token. By doing this, etcd can generate unique cluster IDs and member IDs for the clusters even if they otherwise have the exact same configuration. This can protect etcd from cross-cluster-interaction, which might corrupt the clusters.

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md (hunk 18 above)
Comment (chulkilee) [reply docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: It seems like this etcd ca key is only required when adding new kube master with etcd running in it. If so I agree that once the ca key is not needed, kubeadmn advise users to remove it from the server and keep the copy in the safe place.

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md (hunk 17 above)
Comment (chuckha) [reply docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: I really like this idea. I'll add a section.

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md (hunk 16 above)
Comment (chuckha) [reply docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: I was trying to avoid ordering the nodes, but in practice I think you're right. People will use the first node they ssh to.

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md (hunk 18 above)
Comment (chuckha) [reply docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: I waffled on this bit as well. At first I had it this way, but this required a ton of instructions for moving files around and changing permissions. I ended up going this route to reduce the number of copy instructions. I also received feedback to actually provide a list of commands to move files around, hence the original requirement for SSH between nodes.

Unfortunately for me, I think you're right that I *actually* want to distribute the certs after generating them all on one host. I will have to re-write the workflow.

File: docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md (hunk 22 above)
Comment (chuckha) [reply docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md]: hah yes, that is definitely an easier approach.

**Comment by chuckha** [comment]
oh right. jekyll got switched out for hugo. Also updating that.

**Comment by chuckha** [comment]
@luxas I changed the content significantly and added new content taking your suggestions into account. I'm better about certs & security here.
